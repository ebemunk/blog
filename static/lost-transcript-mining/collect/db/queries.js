export function totalLinesByChar() {
	return `
		select char_name, count(*) as lines
		from dialog
		where type='dialog' and char_name is not null
		group by char_name
		order by lines desc
	;`
}

export function allLines(char_name) {
	return `
		select string_agg(line, E'\n' order by season, episode, seq) as text
		from dialog
		where type='dialog' and char_name='${char_name}'
		group by char_name
	;`
}

export function episodeText(season, episode) {
	return `
		select string_agg(line, ' ' order by seq) as text
		from dialog
		where type='dialog' and season=${season} and episode=${episode} and line!=''
		group by season,episode
		order by season,episode
	;`
}

export function allEpisodes() {
	return `
		select distinct season, episode
		from dialog
		order by season, episode asc
	;`
}
