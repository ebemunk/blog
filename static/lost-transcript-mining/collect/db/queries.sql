-- total lines by char
select char_name, count(*) as lines
from dialog
where type='dialog' and char_name is not null
group by char_name
order by lines desc
;

-- total lines per char per episode per season
select season, episode, char_name, count(*) as lines
from dialog
where type='dialog'
group by season, episode, char_name
order by season, episode, lines desc
;

-- char per episode per season
select distinct season, episode, char_name from dialog
where type='dialog'
order by season, episode, char_name asc
;

-- all spoken dialog prefixed with char name by episode
select season, episode, string_agg(char_name || ': ' || line, E'\n' order by seq)
from dialog
where type='dialog'
group by season,episode
order by season,episode
;

-- all dialog for each char ordered chronologically
select char_name, string_agg(line, E'\n' order by season, episode, seq)
from dialog
where type='dialog'
group by char_name
;

-- chars in every scene
select season, episode, act, scene, array_agg(distinct char_name)
from dialog
where type='dialog'
group by season, episode, act, scene
;