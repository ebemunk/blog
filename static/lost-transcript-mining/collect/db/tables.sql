-- dialog table
CREATE TABLE public.dialog
(
  id SERIAL NOT NULL,
  season integer NOT NULL,
  episode integer NOT NULL,
  seq integer NOT NULL,
  act integer NOT NULL,
  scene integer NOT NULL,
  flashback boolean NOT NULL,
  flashsideways boolean NOT NULL,
  type character varying NOT NULL,
  char_name character varying DEFAULT NULL,
  line text DEFAULT NULL,
  directions json DEFAULT '[]'::json,
  raw text DEFAULT NULL
)

-- watson personality insights
CREATE TABLE public.personality
(
  id SERIAL NOT NULL,
  char_name character varying DEFAULT NULL,
  profile json DEFAULT '{}'::json
)

-- watson tone analyzer for all episodes
CREATE TABLE public.tone
(
  season integer NOT NULL,
  episode integer NOT NULL,
  document_tone json DEFAULT '{}'::json,
  sentences_tone json DEFAULT '[]'::json
)
