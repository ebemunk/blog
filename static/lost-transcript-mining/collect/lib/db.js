export function insertObj(obj, table = 'dialog') {
	const cols = Object.keys(obj)
	.join(',')
	const valOrder = Object.keys(obj)
	.map((v, i) => `$${i+1}`)
	.join(',')
	const sql = `
		INSERT INTO ${table}
		(${cols})
		VALUES
		(${valOrder})
	`
	return [
		sql,
		Object.values(obj)
	]
}
