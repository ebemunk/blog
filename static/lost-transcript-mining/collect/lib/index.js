export * from './scrape-fns'
export * from './db'
