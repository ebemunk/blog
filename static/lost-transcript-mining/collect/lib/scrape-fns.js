import axios from 'axios'
import cheerio from 'cheerio'
import _ from 'lodash'
import c from 'chalk'

const log = process.env.NODE_ENV === 'test' ?
	null : _.partial(console.log, c.bgMagenta('scrape-fns'))

export async function getSeasonList() {
	const { data } = await axios.get('http://lostpedia.wikia.com/wiki/Portal:Transcripts')
	const $ = cheerio.load(data)
	const seasons = _.range(1,7).map(season => {
		const episodes = $(`#Season_${season}`)
		.parent()
		.next('ol')
		.children()
		.get()
		.map((el, i) => {
			const link = $('a', el)
			return {
				season,
				ep: i+1,
				name: link.attr('title').replace(' transcript', ''),
				url: link.attr('href')
			}
		})
		return {
			season,
			episodes
		}
	})
	return seasons
}

export async function getEpisode(episode) {
	const { data } = await axios.get(`http://lostpedia.wikia.com${episode.url}`)
	const $ = cheerio.load(data)
	const lines = $('#toc')
	.nextAll()
	.filter((i, el) => {
		return ['hr', 'h2', 'p'].includes(el.tagName) && $(el).text().trim()
	})
	.map((i, el) => {
		const row = {
			season: episode.season,
			episode: episode.ep,
			// seq: i,
		}
		switch( el.tagName ) {
			case 'hr': return {
				...row,
				type: 'scene'
			}
			case 'h2': return {
				...row,
				type: 'act'
			}
			case 'p': {
				const text = $(el).text().trim()
				row.raw = text
				const stageDirection = text.match(/^\[(.+)\]$/)
				if( stageDirection ) {
					return {
						...row,
						line: stageDirection[1],
						type: 'stageDirection'
					}
				} else {
					const charLine = text.match(/^(.+?)\: (.+)$/)
					// log(charLine)
					if( ! charLine ) return null
					const char_name = charLine[1]
					let line = charLine[2]
					let directions = line.match(/\[([^\]]+)\]/g)
					if( directions ) {
						directions = directions.map(str => str.replace(/[\[\]]/g, ''))
						line = line
						.replace(/\[[^\]]+\][ ]?/g, '')
					}
					return {
						...row,
						char_name,
						line,
						directions,
						type: 'dialog'
					}
				}
			}
		}
	})
	.filter(row => row)
	.get()
	return lines
	.map((row, i) => {
		return {
			...row,
			seq: i+1
		}
	})
}
