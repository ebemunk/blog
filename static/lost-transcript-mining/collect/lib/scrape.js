import fs from 'fs'

import Promise from 'bluebird'
import axios from 'axios'
import cheerio from 'cheerio'
import _ from 'lodash'

const writeFile = Promise.promisify(fs.writeFile)

export async function parseEpisodeIndex(html) {
	const $ = cheerio.load(html)
	const seasons = _.range(1,7)
	.map(season => {
		const episodes = $(`#Season_${season}`)
		// parsing the cheerio tree
		.parent()
		.next('ol')
		.children()
		.get()
		.map((el, i) => {
			const link = $('a', el)
			// an array of these are returned
			return {
				season,
				episode: i+1,
				name: link.attr('title').replace(' transcript', ''),
				url: link.attr('href')
			}
		})
		return episodes
	})
	// flatten
	return seasons.reduce((prev, cur) => prev.concat(cur))
}

export async function saveEpisodeHtml(url, path) {
	const { data } = await axios.get(`${url}`)
	await Promise.fromCallback(cb => fs.writeFile(path, data, cb))
	return data
}

export function getLines(htmlString) {
	const $ = cheerio.load(htmlString)
	const lines = $('#toc')
	// everything after table of contents
	.nextAll()
	// only hr h2 and p tags with content in them
	.filter((i, el) => {
		return ['hr', 'h2', 'p'].includes(el.tagName) && $(el).text().trim()
	})
	.map((i, el) => {
		return {
			tagName: el.tagName,
			text: $(el).text().trim().replace(/\n/g, '')
		}
	})
	.get()
	.map(lineType)
	.filter(Boolean)
	return lines
}

export function lineType({tagName, text}) {
	switch( tagName ) {
		case 'hr': return {
			type: 'scene'
		}

		case 'h2': return {
			type: 'act'
		}

		case 'p': {
			const stageDirection = text.match(/^\[([^\]]+)\]?(?: (.+))?$/)

			if( stageDirection ) {
				return {
					type: 'stageDirection',
					raw: text,
					line: stageDirection[1]
				}
			} else {
				const parsed = parseLine(text)
				if( ! parsed ) return false
				return {
					type: 'dialog',
					raw: text,
					...parsed
				}
			}
		}
	}
}

export function parseLine(text) {
	let line = text
	// everything inside square brackets or paranthesis is a direction
	let directions = []
	// [ and ]
	.concat(text.match(/\[([^\]]+)\]\.?/g))
	// ( and )
	.concat(text.match(/\(([^\]]+)\)\.?/g))
	// remove falsy
	.filter(Boolean)
	// strip and remove from source text
	.map(str => {
		line = line.replace(str, '').replace('  ', ' ')
		return str.replace(/[\[\]\(\)]/g, '').replace(/\.$/, '')
	})

	const charLine = line.trim().match(/^(.+?)\:\s*(.+)?$/)
	if( ! charLine ) {
		console.log('unparseable line:', text)
		console.log()
		return
	}
	const char_name = charLine[1].trim().toUpperCase()
	line = charLine[2] || ''
	return {
		char_name,
		line,
		directions,
	}
}
