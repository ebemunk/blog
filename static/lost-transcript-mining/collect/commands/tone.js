import R from 'ramda'
import Promise from 'bluebird'
import ProgressBar from 'progress'

import { watson } from '../lib'
import {
	getPool,
	queries
} from '../db'
import {
	insertObj,
	logger
} from '../util'

const log = logger('tone')

export default async function tone() {
	const pool = getPool()
	let progress
	return R.pipeP(
		sql => pool.query(sql),
		R.prop('rows'),
		R.tap(rows => {
			log(`found ${rows.length} episodes`)
			progress = new ProgressBar(':current/:total :bar :eta', rows.length)
		}),
		R.partialRight(Promise.mapSeries, [
			R.pipeP(
				({ season, episode }) => watson.episodeTone(pool, season, episode),
				R.evolve({
					document_tone: JSON.stringify,
					sentences_tone: JSON.stringify
				}),
				R.partialRight(insertObj, ['tone']),
				insert => pool.query(...insert),
				R.tap(() => {
					progress.tick()
				})
			)
		]),
	)(queries.allEpisodes())
}
