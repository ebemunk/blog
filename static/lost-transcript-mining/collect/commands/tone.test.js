import tone from './tone'
import { watson } from '../lib'
import * as db from '../db'

describe('tone', () => {
	it('should get tone analysis for all episodes', async () => {
		jasmine.DEFAULT_TIMEOUT_INTERVAL = 350000000

		watson.episodeTone = jest.fn(() => Promise.resolve({
			season: 8,
			episode: 113,
			document_tone: { derp: 23 },
			sentences_tone: [
				{ herder: 1, derper: 2 },
				{ herder: 2, derper: 3 }
			]
		}))
		const query = jest.fn()
		.mockReturnValueOnce(Promise.resolve({
			rows: [
				{ season: 1, episode: 1 },
				{ season: 1, episode: 2 },
			]
		}))
		.mockReturnValue(Promise.resolve())
		db.getPool = jest.fn(() => ({
			query
		}))

		const r = await tone()

		console.log(r);

		expect(watson.episodeTone.mock.calls).toMatchSnapshot()
		expect(query.mock.calls).toMatchSnapshot()
	})
})
