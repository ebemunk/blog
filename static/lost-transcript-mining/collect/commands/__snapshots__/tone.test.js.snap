// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tone should get tone analysis for all episodes 1`] = `
Array [
  Array [
    Object {
      "query": [Function],
    },
    1,
    1,
  ],
  Array [
    Object {
      "query": [Function],
    },
    1,
    2,
  ],
]
`;

exports[`tone should get tone analysis for all episodes 2`] = `
Array [
  Array [
    "
		select distinct season, episode
		from dialog
		order by season, episode asc
	;",
  ],
  Array [
    "
		INSERT INTO tone
		(season,episode,document_tone,sentences_tone)
		VALUES
		($1,$2,$3,$4)
	",
    Array [
      8,
      113,
      "{\\"derp\\":23}",
      "[{\\"herder\\":1,\\"derper\\":2},{\\"herder\\":2,\\"derper\\":3}]",
    ],
  ],
  Array [
    "
		INSERT INTO tone
		(season,episode,document_tone,sentences_tone)
		VALUES
		($1,$2,$3,$4)
	",
    Array [
      8,
      113,
      "{\\"derp\\":23}",
      "[{\\"herder\\":1,\\"derper\\":2},{\\"herder\\":2,\\"derper\\":3}]",
    ],
  ],
]
`;
